instrumentation/github.com/aws/aws-lambda-go/otellambda/lambda.go

```diff
@@ -0,0 +1,351 @@
+func validateArguments(handler reflect.Type) (bool, error) {
+	handlerTakesContext := false
+	if handler.NumIn() > 2 {
+		return false, fmt.Errorf("handlers may not take more than two arguments, but handler takes %d", handler.NumIn())
+	} else if handler.NumIn() > 0 {
+		contextType := reflect.TypeOf((*context.Context)(nil)).Elem()
+		argumentType := handler.In(0)
+		handlerTakesContext = argumentType.Implements(contextType)
+		if handler.NumIn() > 1 && !handlerTakesContext {
+			return false, fmt.Errorf("handler takes two arguments, but the first is not Context. got %s", argumentType.Kind())
+		}
+	}
+
+	return handlerTakesContext, nil
+}
+
+// Ensure handler returns 0-2 values, with an error
+// as its first value if any exist
+func validateReturns(handler reflect.Type) error {
+	errorType := reflect.TypeOf((*error)(nil)).Elem()
+
+	switch n := handler.NumOut(); {
+	case n > 2:
+		return fmt.Errorf("handler may not return more than two values")
+	case n > 1:
+		if !handler.Out(1).Implements(errorType) {
+			return fmt.Errorf("handler returns two values, but the second does not implement error")
+		}
+	case n == 1:
+		if !handler.Out(0).Implements(errorType) {
+			return fmt.Errorf("handler returns a single value, but it does not implement error")
+		}
+	}
+
+	return nil
+}
+
+// Wraps and calls customer lambda handler then unpacks response as necessary
+func wrapperInternals(handlerFunc interface{}, event reflect.Value, ctx context.Context, takesContext bool) (interface{}, error) {
+	wrappedLambdaHandler := reflect.ValueOf(wrapper(handlerFunc))
+
+	argsWrapped := []reflect.Value{reflect.ValueOf(ctx), event, reflect.ValueOf(takesContext)}
+	response := wrappedLambdaHandler.Call(argsWrapped)[0].Interface().([]reflect.Value)
+
+	// convert return values into (interface{}, error)
+	var err error
+	if len(response) > 0 {
+		if errVal, ok := response[len(response)-1].Interface().(error); ok {
+			err = errVal
+		}
+	}
+	var val interface{}
+	if len(response) > 1 {
+		val = response[0].Interface()
+	}
+
+	return val, err
+}
+
+// converts the given payload to the correct event type
+func payloadToEvent(eventType reflect.Type, payload interface{}) (reflect.Value, error) {
+	event := reflect.New(eventType)
+
+	// lambda SDK normally unmarshalls to customer event type, however
+	// with the wrapper the SDK unmarshalls to map[string]interface{}
+	// due to our use of reflection. Therefore we must convert this map
+	// to customer's desired event, we do so by simply re-marshalling then
+	// unmarshalling to the desired event type
+	remarshalledPayload, err := json.Marshal(payload)
+	if err != nil {
+		return reflect.Value{}, err
+	}
+
+	if err := json.Unmarshal(remarshalledPayload, event.Interface()); err != nil {
+		return reflect.Value{}, err
+	}
+	return event, nil
+}
+
+// LambdaHandlerWrapper Provides a lambda handler which wraps customer lambda handler with OTel Tracing
+func LambdaHandlerWrapper(handlerFunc interface{}) interface{} {
```

Comment:
```suggestion
func WrapHandler(handlerFunc interface{}) interface{} {
```

`Lambda` can be dropped since it will be in the package name when referenced externally, and using a verb phrase rather than a noun for naming a function makes more sense to me.  I'd expect a `HandlerWrapper` to be a struct or interface rather than a function.

Reply:
Renamed wrappers to `WrapHandlerFunction` and `WrapHandler`.